lib/slurm_squeue_client.rb

```diff
@@ -61,7 +61,73 @@ def sinfo
     
     s.success? ? @sinfo = o : raise(CommandFailed, e)
   end
-  
+
+  # Return length of GRES field from SLURM
+  # @return [Integer] gres length
+  def gres_length
+    return @gres_length if defined?(@gres_length)
+
+    Open3.pipeline_rw "sinfo -o '%G' | awk '{ print length }'", "sort -n", "tail -1" do |stdin,stdout|
+      stdin.write stdout
+      stdin.close
+      @gres_length = stdout.read.to_i
+    end
+  end
+
+  # Parse and return total number of GPU nodes in a SLURM cluster
+  # @return [Integer] number of GPU nodes
+  def gpu_nodes
+    return @available_gpu_nodes if defined?(@available_gpu_nodes)
+
+    Open3.pipeline_rw "sinfo -N -h -a --Format='nodehost,gres:#{@gres_length}'", 'uniq', 'grep gpu:', 'wc -l' do |stdin, stdout|
+      stdin.write stdout
+      stdin.close
+      @available_gpu_nodes = stdout.read.to_i
+    end
+  end
+
+  # Return number of GPU nodes with mixed or idle status
+  # 
+  # @return [Integer] number of GPU nodes with status mixed (some CPUs allocated)
+  def gpu_nodes_free
+    return @gpu_nodes_free if defined?(@gpu_nodes_free)
+
+    Open3.pipeline_rw "sinfo -a -h --Node --Format='nodehost,gres:#{@gres_length}',statelong", 'uniq', 'grep gpu:', 'egrep "idle"', 'wc -l' do |stdin, stdout|
+      stdin.write stdout
+      stdin.close
+      @gpu_nodes_free = stdout.read.to_i
+    end
+  end
+
+  # Return number of GPU nodes in use
+  # 
+  # @return [Integer] gpu nodes in use
+  def gpu_nodes_active
+    return @gpu_nodes_active if defined?(@gpu_nodes_active)
+
+    @gpu_nodes_active = @available_gpu_nodes - @gpu_nodes_free
+  end
+
+  # Returns percentage of GPU nodes that are available
+  # 
+  # @return [Float] percentage gpu nodes available
+  def gpu_nodes_available_percent
+    ((gpu_nodes - gpu_nodes_free).to_f / gpu_nodes.to_f) * 100
+  end
+
+  # Number of pending jobs requesting GPUs
+  # 
+  # @return [Integer] number of pending jobs requesting GPUs
+  def gpu_jobs_pending
+    return @gpu_jobs_pending if defined?(@gpu_jobs_pending)
+
+    Open3.pipeline_rw "squeue --states=PENDING -O 'jobid,tres-per-job:#{gres_length},tres-per-node:#{gres_length},tres-per-socket:#{gres_length},tres-per-task:#{gres_length}' -h", "grep gpu:", "wc -l" do |stdin, stdout|
+      stdin.write stdout
+      stdin.close
```

Comment:
In this case it seems like you would want to use `Open3.pipeline_r` instead of `Open3.pipeline_rw` as you don't need to write anything to stdin of the sinfo command, you just need to read stdout.